File: packages/ui-compose/ui/src/main/kotlin/com/kinetixui/ui/Spinner.kt
package com.kinetixui.ui

import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.rotate
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.res.dimensionResource
import androidx.compose.ui.unit.Dp

/**
 * KinetixSpinner — mirrors `packages/ui/src/components/spinner.tsx`
 * (a `border-2 border-current border-t-transparent` ring, `animate-spin`).
 * Compose has no CSS-current-color / border-side-transparent equivalent, so
 * this draws the same shape directly: a 270°-sweep arc (i.e. one quadrant
 * left transparent, matching `border-t-transparent`), rotating continuously.
 */
enum class KinetixSpinnerSize { Sm, Md, Lg }
enum class KinetixSpinnerVariant { Default, Muted, OnColor }

@Composable
fun KinetixSpinner(
    modifier: Modifier = Modifier,
    size: KinetixSpinnerSize = KinetixSpinnerSize.Md,
    variant: KinetixSpinnerVariant = KinetixSpinnerVariant.Default,
) {
    val colors = KinetixColorScheme.current
    val color = when (variant) {
        KinetixSpinnerVariant.Default -> colors.primary
        KinetixSpinnerVariant.Muted -> colors.mutedForeground
        KinetixSpinnerVariant.OnColor -> colors.primaryForeground
    }
    val diameter: Dp = when (size) {
        KinetixSpinnerSize.Sm -> dimensionResource(R.dimen.spacing_4) // 16dp
        KinetixSpinnerSize.Md -> dimensionResource(R.dimen.spacing_6) // 24dp
        KinetixSpinnerSize.Lg -> dimensionResource(R.dimen.spacing_8) // 32dp
    }
    val strokeWidth = dimensionResource(R.dimen.border_width_focus) // 2dp, matches `border-2`

    val transition = rememberInfiniteTransition(label = "KinetixSpinner")
    val angle by transition.animateFloat(
        initialValue = 0f,
        targetValue = 360f,
        animationSpec = infiniteRepeatable(tween(durationMillis = 800, easing = LinearEasing)),
        label = "KinetixSpinnerAngle",
    )

    Canvas(modifier = modifier.size(diameter).rotate(angle)) {
        drawArc(
            color = color,
            startAngle = 0f,
            sweepAngle = 270f, // leaves one quadrant open, i.e. `border-t-transparent`
            useCenter = false,
            style = Stroke(width = strokeWidth.toPx()),
        )
    }
}

File: packages/ui-swiftui/Sources/KinetixUI/Spinner.swift
//
// Spinner.swift — KinetixSpinner.
//
// Mirrors packages/ui/src/components/spinner.tsx (a `border-2
// border-current border-t-transparent` ring, `animate-spin`). SwiftUI has
// no border-side-transparent equivalent, so this draws the shape
// directly: a 270°-sweep stroked circle (one quadrant left open, i.e.
// `border-t-transparent`), rotating continuously — same approach as the
// Compose port.
//

import SwiftUI

public enum KinetixSpinnerSize {
    case sm, md, lg
}

public enum KinetixSpinnerVariant {
    case `default`, muted, onColor
}

public struct KinetixSpinner: View {
    @Environment(\.kinetixColors) private var colors
    @State private var spinning = false

    private let size: KinetixSpinnerSize
    private let variant: KinetixSpinnerVariant

    public init(size: KinetixSpinnerSize = .md, variant: KinetixSpinnerVariant = .default) {
        self.size = size
        self.variant = variant
    }

    private var diameter: CGFloat {
        switch size {
        case .sm: return 16 // size-4
        case .md: return 24 // size-6
        case .lg: return 32 // size-8
        }
    }

    private var color: Color {
        switch variant {
        case .default: return colors.primary
        case .muted:   return colors.mutedForeground
        case .onColor: return colors.primaryForeground
        }
    }

    public var body: some View {
        Circle()
            .trim(from: 0, to: 0.75) // border-t-transparent → one quadrant open
            .stroke(color, lineWidth: 2) // border-2
            .frame(width: diameter, height: diameter)
            .rotationEffect(.degrees(spinning ? 360 : 0))
            .animation(.linear(duration: 0.8).repeatForever(autoreverses: false), value: spinning)
            .onAppear { spinning = true }
            .accessibilityLabel("Loading")
    }
}
